// iamport-ionic4-kcp/esm2015/lib/iamport.certification.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/iamport.certification.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
export class IamportCertification {
    /**
     * @param {?} response
     */
    constructor(response) {
        this.success = response.success;
        this.impUid = response.imp_uid;
        this.merchantUid = response.merchant_uid;
    }
    /**
     * @return {?}
     */
    isSuccess() {
        return this.success;
    }
    /**
     * @return {?}
     */
    getImpUid() {
        return this.impUid;
    }
    /**
     * @return {?}
     */
    getMerchantUid() {
        return this.merchantUid;
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.success;
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.impUid;
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.merchantUid;
}
