// iamport-ionic4-kcp/esm5/lib/iamport.payment.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/iamport.payment.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var IamportPayment = /** @class */ (function () {
    function IamportPayment(response) {
        this.success = response.success;
        this.status = response.status;
        this.response = response;
    }
    /**
     * @return {?}
     */
    IamportPayment.prototype.isSuccess = /**
     * @return {?}
     */
    function () {
        return this.success;
    };
    /**
     * @return {?}
     */
    IamportPayment.prototype.getStatus = /**
     * @return {?}
     */
    function () {
        return this.status;
    };
    /**
     * @return {?}
     */
    IamportPayment.prototype.getResponse = /**
     * @return {?}
     */
    function () {
        return this.response;
    };
    return IamportPayment;
}());
export { IamportPayment };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IamportPayment.prototype.success;
    /**
     * @type {?}
     * @private
     */
    IamportPayment.prototype.status;
    /**
     * @type {?}
     * @private
     */
    IamportPayment.prototype.response;
}
