// iamport-ionic4-kcp/esm5/lib/iamport.certification.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/iamport.certification.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
var IamportCertification = /** @class */ (function () {
    function IamportCertification(response) {
        this.success = response.success;
        this.impUid = response.imp_uid;
        this.merchantUid = response.merchant_uid;
    }
    /**
     * @return {?}
     */
    IamportCertification.prototype.isSuccess = /**
     * @return {?}
     */
    function () {
        return this.success;
    };
    /**
     * @return {?}
     */
    IamportCertification.prototype.getImpUid = /**
     * @return {?}
     */
    function () {
        return this.impUid;
    };
    /**
     * @return {?}
     */
    IamportCertification.prototype.getMerchantUid = /**
     * @return {?}
     */
    function () {
        return this.merchantUid;
    };
    return IamportCertification;
}());
export { IamportCertification };
if (false) {
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.success;
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.impUid;
    /**
     * @type {?}
     * @private
     */
    IamportCertification.prototype.merchantUid;
}
